File: src/rust/src/catalog.rs
//! Catalog connections.
//!
//! Three catalog kinds are reachable from R: `rest`, `memory` and `glue`.
//! `memory` is an in-process catalog over a warehouse directory, which is what
//! makes fully offline use -- and the entire test suite -- possible.
//!
//! Note that iceberg-rust has no Hadoop or filesystem catalog, so there is
//! deliberately no `hadoop` option here; `memory` is the local-warehouse
//! equivalent.
//!
//! Catalog properties are passed straight through as an opaque map. Their values
//! frequently are credentials, so no code path here formats them into a
//! message, a log line or a print method.

use std::collections::HashMap;
use std::sync::Arc;

use extendr_api::prelude::*;
use iceberg::io::{LocalFsStorageFactory, StorageFactory};
use iceberg::memory::MemoryCatalogBuilder;
use iceberg::{Catalog, CatalogBuilder, NamespaceIdent};
use iceberg_catalog_rest::RestCatalogBuilder;

// not_compiled_in is referenced by full path below: whether it is used at all
// depends on which optional features are enabled, and an import that is unused
// under some feature combinations is a warning we would rather not have.
use crate::errors::{RResult, config_err, ctx};
use crate::runtime::{block_on, iceberg_runtime};

/// A live catalog connection, handed to R as an external pointer.
///
/// Only the catalog itself. The kind and the connection name live on the R-side
/// `icebergr_catalog` object, which is the one place they are read from; keeping
/// a second copy here that nothing reads is state that can only drift.
pub struct RCatalog {
    pub inner: Arc<dyn Catalog>,
}

/// Whether `w` starts with `scheme://`, ignoring case.
///
/// A URI scheme is case-insensitive, and iceberg-rust's `FileIO` treats it so,
/// since it parses locations as URLs: `S3://bucket` is an S3 location to it. The
/// R side already matched `s3://` case-insensitively when deciding whether to
/// forward AWS credentials, so a case-sensitive test here disagreed with it and
/// put an `S3://` warehouse on the local filesystem.
fn has_scheme(w: &str, scheme: &str) -> bool {
    w.len() > scheme.len() + 3
        && w.as_bytes()[..scheme.len()].eq_ignore_ascii_case(scheme.as_bytes())
        && w.as_bytes()[scheme.len()..].starts_with(b"://")
}

fn looks_like_local_path(w: &str) -> bool {
    let b = w.as_bytes();
    has_scheme(w, "file")
        || w.starts_with('/')
        // Windows drive letter, e.g. C:\warehouse. The leading byte has to be a
        // letter: without that check any string with a colon in second position
        // is read as a local path.
        || (b.len() > 2
            && b[0].is_ascii_alphabetic()
            && b[1] == b':'
            && (b[2] == b'\\' || b[2] == b'/'))
}

#[cfg(feature = "s3")]
fn s3_factory() -> RResult<Arc<dyn StorageFactory>> {
    Ok(Arc::new(
        iceberg_storage_opendal::OpenDalStorageFactory::S3 {
            customized_credential_load: None,
        },
    ))
}

#[cfg(not(feature = "s3"))]
fn s3_factory() -> RResult<Arc<dyn StorageFactory>> {
    Err(crate::errors::not_compiled_in(
        "Object storage (S3) access",
        "s3",
    ))
}

/// Choose the storage backend.
///
/// `None` means "leave the builder's own default alone", which is the right
/// answer for a REST catalog that reports its own storage configuration.
fn storage_factory(
    storage: &str,
    warehouse: Option<&str>,
) -> RResult<Option<Arc<dyn StorageFactory>>> {
    match resolve_storage(storage, warehouse) {
        "local" => Ok(Some(
            Arc::new(LocalFsStorageFactory) as Arc<dyn StorageFactory>
        )),
        "s3" => s3_factory().map(Some),
        "default" => Ok(None),
        other => Err(extendr_api::Error::Other(format!(
            "unknown storage backend {other:?}; expected \"auto\", \"local\" or \"s3\""
        ))),
    }
}

/// Which backend `storage` names, reading `"auto"` off the warehouse.
fn resolve_storage<'a>(storage: &'a str, warehouse: Option<&str>) -> &'a str {
    if storage != "auto" {
        return storage;
    }
    match warehouse {
        Some(w) if has_scheme(w, "s3") || has_scheme(w, "s3a") => "s3",
        Some(w) if looks_like_local_path(w) => "local",
        _ => "default",
    }
}

#[cfg(feature = "glue")]
fn connect_glue(
    name: &str,
    props: HashMap<String, String>,
    factory: Option<Arc<dyn StorageFactory>>,
    keys: &[String],
) -> RResult<Arc<dyn Catalog>> {
    let mut builder =
        iceberg_catalog_glue::GlueCatalogBuilder::default().with_runtime(iceberg_runtime());
    // Glue tables live in object storage, so default to S3 rather than to the
    // local filesystem.
    let factory = match factory {
        Some(f) => f,
        None => s3_factory()?,
    };
    builder = builder.with_storage_factory(factory);

    let catalog = block_on(builder.load(name.to_string(), props))
        .map_err(|e| config_err("could not connect to the AWS Glue catalog", keys, e))?;
    Ok(Arc::new(catalog))
}

#[cfg(not(feature = "glue"))]
fn connect_glue(
    _name: &str,
    _props: HashMap<String, String>,
    _factory: Option<Arc<dyn StorageFactory>>,
    _keys: &[String],
) -> RResult<Arc<dyn Catalog>> {
    Err(crate::errors::not_compiled_in(
        "The AWS Glue catalog",
        "glue",
    ))
}

#[extendr]
fn rs_catalog_connect(
    kind: &str,
    name: &str,
    storage: &str,
    keys: Vec<String>,
    values: Vec<String>,
) -> RResult<ExternalPtr<RCatalog>> {
    if keys.len() != values.len() {
        return Err("internal error: property keys and values differ in length".into());
    }

    let props: HashMap<String, String> = keys.iter().cloned().zip(values).collect();
    let warehouse = props.get("warehouse").cloned();
    let factory = storage_factory(storage, warehouse.as_deref())?;

    let inner: Arc<dyn Catalog> = match kind {
        "memory" => {
            // MemoryCatalog defaults to *in-memory* storage, which would write a
            // local warehouse into oblivion, so a factory is always set.
            let factory = factory.unwrap_or_else(|| Arc::new(LocalFsStorageFactory));
            let builder = MemoryCatalogBuilder::default()
                .with_runtime(iceberg_runtime())
                .with_storage_factory(factory);
            let catalog = block_on(builder.load(name.to_string(), props))
                .map_err(|e| config_err("could not open the memory catalog", &keys, e))?;
            Arc::new(catalog)
        }
        "rest" => {
            let mut builder = RestCatalogBuilder::default().with_runtime(iceberg_runtime());
            if let Some(f) = factory {
                builder = builder.with_storage_factory(f);
            }
            let catalog = block_on(builder.load(name.to_string(), props))
                .map_err(|e| config_err("could not connect to the REST catalog", &keys, e))?;
            Arc::new(catalog)
        }
        "glue" => connect_glue(name, props, factory, &keys)?,
        other => {
            return Err(extendr_api::Error::Other(format!(
                "unknown catalog type {other:?}.\n\
                 icebergr supports \"rest\", \"memory\" and \"glue\". Note that \
                 iceberg-rust has no Hadoop or filesystem catalog; use \
                 type = \"memory\" with a warehouse directory for a local table."
            )));
        }
    };

    Ok(ExternalPtr::new(RCatalog { inner }))
}

/// Namespaces are multi-level; each is returned as a dot-joined string.
#[extendr]
fn rs_list_namespaces(cat: ExternalPtr<RCatalog>, parent: Vec<String>) -> RResult<Vec<String>> {
    let parent = if parent.is_empty() {
        None
    } else {
        Some(NamespaceIdent::from_vec(parent).map_err(|e| ctx("invalid namespace", e))?)
    };

    let found = block_on(cat.inner.list_namespaces(parent.as_ref()))
        .map_err(|e| ctx("could not list namespaces", e))?;

    Ok(found.iter().map(|n| n.as_ref().join(".")).collect())
}

/// Table names within `namespace`, without the namespace prefix.
#[extendr]
fn rs_list_tables(cat: ExternalPtr<RCatalog>, namespace: Vec<String>) -> RResult<Vec<String>> {
    let ns = NamespaceIdent::from_vec(namespace).map_err(|e| ctx("invalid namespace", e))?;
    let found =
        block_on(cat.inner.list_tables(&ns)).map_err(|e| ctx("could not list tables", e))?;
    Ok(found.iter().map(|t| t.name().to_string()).collect())
}

#[extendr]
fn rs_namespace_exists(cat: ExternalPtr<RCatalog>, namespace: Vec<String>) -> RResult<bool> {
    let ns = NamespaceIdent::from_vec(namespace).map_err(|e| ctx("invalid namespace", e))?;
    block_on(cat.inner.namespace_exists(&ns)).map_err(|e| ctx("could not check namespace", e))
}

#[extendr]
fn rs_create_namespace(cat: ExternalPtr<RCatalog>, namespace: Vec<String>) -> RResult<()> {
    let ns = NamespaceIdent::from_vec(namespace).map_err(|e| ctx("invalid namespace", e))?;
    block_on(cat.inner.create_namespace(&ns, HashMap::new()))
        .map_err(|e| ctx("could not create namespace", e))?;
    Ok(())
}

extendr_module! {
    mod catalog;
    fn rs_catalog_connect;
    fn rs_list_namespaces;
    fn rs_list_tables;
    fn rs_namespace_exists;
    fn rs_create_namespace;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_scheme_matches_in_any_case_and_only_as_a_scheme() {
        assert!(has_scheme("s3://bucket/wh", "s3"));
        assert!(has_scheme("S3://bucket/wh", "s3"));
        assert!(has_scheme("File:///data/wh", "file"));
        // A prefix of the scheme, or the scheme with nothing after it, is not.
        assert!(!has_scheme("s3a://bucket/wh", "s3"));
        assert!(!has_scheme("s3://", "s3"));
        assert!(!has_scheme("s3", "s3"));
        assert!(!has_scheme("", "s3"));
    }

    #[test]
    fn auto_reads_the_backend_off_the_warehouse() {
        // The case that disagreed with the R side: an upper-case scheme was not
        // object storage here, so a memory catalog put it on the local disk.
        assert_eq!(resolve_storage("auto", Some("S3://bucket/wh")), "s3");
        assert_eq!(resolve_storage("auto", Some("s3a://bucket/wh")), "s3");
        assert_eq!(resolve_storage("auto", Some("/data/wh")), "local");
        assert_eq!(resolve_storage("auto", Some("C:/data/wh")), "local");
        assert_eq!(resolve_storage("auto", Some("FILE:///data/wh")), "local");
        // A REST warehouse identified by name leaves the builder's default.
        assert_eq!(resolve_storage("auto", Some("analytics")), "default");
        assert_eq!(resolve_storage("auto", None), "default");
        // An explicit choice is never second-guessed.
        assert_eq!(resolve_storage("local", Some("s3://bucket/wh")), "local");
    }
}

File: R/catalog.R
# Catalog connections.
#
# Credentials are read from the environment and never taken as arguments. An
# argument holding a bearer token ends up in the script that called it, in
# .Rhistory, and in any knitr cache of the chunk that ran it; an environment
# variable does not.

# Environment variable -> Iceberg catalog property.
credential_vars <- c(
  ICEBERGR_REST_TOKEN = "token",
  ICEBERGR_REST_CREDENTIAL = "credential",
  ICEBERGR_REST_OAUTH2_SERVER_URI = "oauth2-server-uri",
  ICEBERGR_REST_SCOPE = "scope",
  ICEBERGR_S3_ACCESS_KEY_ID = "s3.access-key-id",
  ICEBERGR_S3_SECRET_ACCESS_KEY = "s3.secret-access-key",
  ICEBERGR_S3_SESSION_TOKEN = "s3.session-token",
  ICEBERGR_S3_REGION = "s3.region",
  ICEBERGR_S3_ENDPOINT = "s3.endpoint"
)

# Standard AWS variables, consulted when the icebergr-specific ones are unset so
# that an already-configured AWS environment simply works.
aws_fallbacks <- c(
  AWS_ACCESS_KEY_ID = "s3.access-key-id",
  AWS_SECRET_ACCESS_KEY = "s3.secret-access-key",
  AWS_SESSION_TOKEN = "s3.session-token",
  AWS_REGION = "s3.region",
  AWS_DEFAULT_REGION = "s3.region"
)

# Property names that should never be typed into a script.
secret_props <- c(
  "token", "credential", "s3.access-key-id", "s3.secret-access-key",
  "s3.session-token", "gcs.oauth2.token", "adls.sas-token",
  "adls.connection-string"
)

# Does this connection actually address object storage?
#
# `storage = "s3"` says so outright, Glue implies it, and an `s3://` warehouse
# is the "auto" case. A REST catalog whose warehouse is a *name* rather than a
# location cannot be classified from here, which is what `storage = "s3"` is
# for.
#' @noRd
uses_object_storage <- function(type, storage, warehouse) {
  if (identical(storage, "s3")) {
    return(TRUE)
  }
  if (identical(storage, "local")) {
    return(FALSE)
  }
  if (identical(type, "glue")) {
    return(TRUE)
  }
  !is.null(warehouse) && grepl("^s3a?://", warehouse, ignore.case = TRUE)
}

# The ICEBERGR_* variables are an explicit instruction and are always honoured.
# The ambient AWS_* ones are not: they are whatever the machine happens to have
# configured, and forwarding them to a catalog that is not object storage hands
# the user's keys to a party with no use for them. Worse, a third-party REST
# catalog controls the table `location` and can answer with its own
# `s3.endpoint`, so the client would then sign SigV4 requests to a
# server-nominated host using those keys. Scoped to the connections that can
# actually need them.
#' @noRd
properties_from_env <- function(type = "rest", storage = "auto", warehouse = NULL) {
  out <- list()

  for (var in names(credential_vars)) {
    value <- Sys.getenv(var, unset = "")
    if (nzchar(value)) out[[credential_vars[[var]]]] <- value
  }

  if (uses_object_storage(type, storage, warehouse)) {
    for (var in names(aws_fallbacks)) {
      prop <- aws_fallbacks[[var]]
      if (!is.null(out[[prop]])) next
      value <- Sys.getenv(var, unset = "")
      if (nzchar(value)) out[[prop]] <- value
    }
  }

  out
}

# The host part of a URI, with any userinfo and port removed.
#' @noRd
uri_host <- function(x) {
  authority <- sub("[/?#].*$", "", sub("^[^:]*://", "", x))
  authority <- sub("^.*@", "", authority)
  sub(":[0-9]+$", "", authority)
}

# A URI that would carry a credential in the clear.
#
# No scheme at all is not this function's business -- iceberg-rust will reject
# it -- and loopback over http is the ordinary development case.
#' @noRd
sends_in_cleartext <- function(x) {
  if (is.null(x) || !nzchar(x) || !grepl("://", x, fixed = TRUE)) {
    return(FALSE)
  }
  scheme <- tolower(sub("://.*$", "", x))
  if (scheme %in% c("https", "wss")) {
    return(FALSE)
  }
  !(uri_host(x) %in% c("localhost", "127.0.0.1", "::1", "[::1]", "0.0.0.0"))
}

# Refuse to put a credential on the wire unencrypted.
#
# `properties_from_env()` has already loaded whatever the environment holds, so
# by this point a bearer token is one `block_on` away from an `Authorization`
# header. Over `http://` that header, and the OAuth2 exchange's
# `client_id:client_secret` body, are readable by anything on the path. An error
# rather than a warning, because a warning does not stop the request; the escape
# hatch is one environment variable, and loopback is exempt already.
#' @noRd
check_credential_transport <- function(props, call = rlang::caller_env()) {
  supplied <- intersect(names(props), secret_props)
  if (!length(supplied)) {
    return(invisible(NULL))
  }
  if (env_flag("ICEBERGR_ALLOW_INSECURE_CREDENTIALS")) {
    return(invisible(NULL))
  }

  endpoints <- c(uri = props[["uri"]], `oauth2-server-uri` = props[["oauth2-server-uri"]])
  bad <- endpoints[vapply(endpoints, sends_in_cleartext, logical(1))]
  if (!length(bad)) {
    return(invisible(NULL))
  }

  abort(
    c(
      paste0(
        "Refusing to send a credential to ",
        encodeString(unname(bad)[[1]]), " over an unencrypted connection."
      ),
      i = paste0(
        "These credential properties are set: ",
        paste(sort(supplied), collapse = ", "), "."
      ),
      i = "Use an https:// endpoint, or unset the credential.",
      i = paste0(
        "To proceed anyway, set ICEBERGR_ALLOW_INSECURE_CREDENTIALS=true. ",
        "A loopback address needs no exemption."
      )
    ),
    call = call
  )
}

#' Connect to an Iceberg catalog
#'
#' @param type Catalog type. `"rest"` for an Iceberg REST catalog, `"memory"` for
#'   an in-process catalog over a local warehouse directory, `"glue"` for AWS
#'   Glue.
#'
#'   There is deliberately no `"hadoop"` option: `iceberg-rust` does not
#'   implement a Hadoop or filesystem catalog. Use `type = "memory"` with
#'   `warehouse` for a table on local disk.
#' @param uri Catalog URI. Required for `type = "rest"`, ignored otherwise.
#' @param warehouse Warehouse location. A directory for `type = "memory"`; for
#'   REST catalogs, the warehouse name or location the server expects.
#' @param ... Further catalog properties, passed through to `iceberg-rust` as
#'   name-value pairs. Use this for non-secret configuration such as
#'   `"s3.endpoint"` or `"rest.signing-region"`.
#' @param storage Storage backend. `"auto"` infers it from `warehouse`,
#'   `"local"` forces the local filesystem, `"s3"` forces object storage. S3
#'   requires the package to have been compiled with the `s3` Cargo feature.
#' @param name A label for the connection, used in error messages.
#'
#' @section Credentials:
#' Credentials are read from environment variables, never from arguments:
#'
#' \describe{
#'   \item{`ICEBERGR_REST_TOKEN`}{Bearer token for a REST catalog.}
#'   \item{`ICEBERGR_REST_CREDENTIAL`}{OAuth2 client credential.}
#'   \item{`ICEBERGR_REST_OAUTH2_SERVER_URI`}{OAuth2 token endpoint.}
#'   \item{`ICEBERGR_REST_SCOPE`}{OAuth2 scope.}
#'   \item{`ICEBERGR_S3_ACCESS_KEY_ID`, `ICEBERGR_S3_SECRET_ACCESS_KEY`,
#'     `ICEBERGR_S3_SESSION_TOKEN`}{Object storage credentials.}
#' }
#'
#' The standard `AWS_*` variables are a fallback for the `ICEBERGR_S3_*` ones,
#' but only for a connection that addresses object storage: `storage = "s3"`,
#' `type = "glue"`, or an `s3://` `warehouse`. They are *not* forwarded to a
#' catalog that has no object storage in sight, because a third-party REST
#' catalog controls each table's `location` and may answer with its own
#' `s3.endpoint` -- at which point ambient keys would sign requests to a host it
#' chose. Set `storage = "s3"` if a REST catalog identified by name needs them.
#'
#' A credential is never sent over an unencrypted connection: an `http://`
#' `uri` or OAuth2 endpoint is an error whenever any credential property is
#' populated. A loopback address is exempt, since developing against a local
#' catalog is ordinary, and `ICEBERGR_ALLOW_INSECURE_CREDENTIALS=true` overrides
#' the check.
#'
#' Catalog properties are never printed, logged or included in error messages,
#' and `user:password@` in a `uri` is redacted when a catalog is printed. A
#' credential property passed through `...` anyway is accepted but warned about,
#' since a script is the one place it should not be.
#'
#' @return An `icebergr_catalog` object.
#'
#' @examples
#' # A local warehouse needs no catalog server and no credentials.
#' warehouse <- tempfile("warehouse")
#' dir.create(warehouse)
#' catalog <- icebergr_catalog("memory", warehouse = warehouse)
#' catalog
#'
#' \dontrun{
#' # A REST catalog. The token comes from the environment, not from here.
#' Sys.setenv(ICEBERGR_REST_TOKEN = "...")
#' catalog <- icebergr_catalog("rest", uri = "https://catalog.example.com")
#' }
#' @export
icebergr_catalog <- function(type = c("rest", "memory", "glue"),
                             uri = NULL,
                             warehouse = NULL,
                             ...,
                             storage = c("auto", "local", "s3"),
                             name = "icebergr") {
  ensure_rust()

  type <- match.arg(type)
  storage <- match.arg(storage)
  check_string(uri, "uri")
  check_string(warehouse, "warehouse")
  check_string(name, "name", allow_null = FALSE)

  extra <- list(...)
  if (length(extra) && (is.null(names(extra)) || any(!nzchar(names(extra))))) {
    abort("All catalog properties passed through `...` must be named.")
  }

  # Blank counts as absent. `uri = Sys.getenv("MY_CATALOG")` returns "" when the
  # variable is unset, which is the common way to arrive here empty-handed, and
  # it deserves the same answer as a missing argument rather than an error from
  # inside iceberg-rust.
  if (type == "rest" && (is.null(uri) || !nzchar(trimws(uri)))) {
    abort(c(
      "`uri` is required for a REST catalog.",
      i = "For example: icebergr_catalog(\"rest\", uri = \"https://catalog.example.com\")"
    ))
  }

  if (type == "memory") {
    if (is.null(warehouse)) {
      abort(c(
        "`warehouse` is required for a memory catalog.",
        i = "It is the directory the table data lives in."
      ))
    }
    # Slash-separated, because this becomes the root of every table location
    # Iceberg writes into the table's own metadata. On Windows normalizePath()
    # returns backslashes, which iceberg-rust then joins its own "/" onto, giving
    # a mixed "C:\warehouse/db/events" that happens to parse here and is a poor
    # thing to hand another engine reading the same warehouse.
    warehouse <- as_iceberg_location(normalizePath(warehouse, mustWork = FALSE))
    if (!dir.exists(warehouse)) {
      abort(paste0(
        "The warehouse directory does not exist: ",
        encodeString(warehouse, quote = "\"")
      ))
    }
  }

  props <- properties_from_env(type, storage, warehouse)
  if (!is.null(uri)) props[["uri"]] <- uri
  if (!is.null(warehouse)) props[["warehouse"]] <- warehouse

  for (key in names(extra)) {
    value <- extra[[key]]
    if (length(value) != 1L || is.na(value)) {
      abort(paste0("Catalog property `", key, "` must be a single non-missing value."))
    }
    if (key %in% secret_props) {
      env_var <- names(credential_vars)[match(key, credential_vars)]
      warn(c(
        paste0("Passing ", encodeString(key, quote = "\""), " as an argument risks leaking it."),
        i = "It will be visible in your script, your .Rhistory and any knitr cache.",
        i = if (!is.na(env_var)) {
          paste0("Set the ", env_var, " environment variable instead.")
        } else {
          "Set it in the environment instead."
        }
      ))
    }
    props[[key]] <- as.character(value)
  }

  # After `...`, so that a credential passed there is covered too.
  check_credential_transport(props)

  # A catalog with no properties at all is legitimate (a REST catalog whose
  # server is configured elsewhere). names() and unlist() both give NULL for an
  # empty list, and Rust wants a character vector, so normalise here.
  ptr <- rs_catalog_connect(
    kind = type,
    name = name,
    storage = storage,
    keys = as.character(names(props)),
    values = as.character(unlist(props, use.names = FALSE))
  )

  structure(
    list(ptr = ptr, type = type, name = name, uri = uri, warehouse = warehouse),
    class = "icebergr_catalog"
  )
}

#' @noRd
check_catalog <- function(x, call = rlang::caller_env()) {
  if (!inherits(x, "icebergr_catalog")) {
    abort("`catalog` must be an object created by `icebergr_catalog()`.", call = call)
  }
  check_live_ptr(x$ptr, "catalog", call = call)
  invisible(NULL)
}

#' @noRd
as_namespace <- function(x, arg = "namespace", allow_null = FALSE,
                         call = rlang::caller_env()) {
  # character(0) means the same as NULL: no levels were named. Letting it past
  # here produced an empty level vector that nothing downstream can use --
  # icebergr_create_namespace() reached `seq_len(-1L)` and reported "argument
  # must be coercible to non-negative integer", which names neither the argument
  # nor the mistake.
  if (is.null(x) || !length(x)) {
    if (allow_null) {
      return(character())
    }
    abort(paste0("`", arg, "` is required."), call = call)
  }
  if (!is.character(x) || anyNA(x) || !all(nzchar(x))) {
    abort(paste0("`", arg, "` must be a character vector of namespace levels."), call = call)
  }
  # "a.b" and c("a", "b") mean the same thing. unlist() of an empty list is
  # NULL rather than character(), which Rust would refuse.
  levels <- as.character(unlist(strsplit(x, ".", fixed = TRUE), use.names = FALSE))

  # A doubled separator splits to an empty level, and an empty level is not
  # something a catalog can address: "a..b" would otherwise create the three
  # levels c("a", "", "b"), which lists back as "a." and can never be named
  # again -- parse_identifier() reads "a..b.events" as c("a", "b"). Dropped
  # rather than rejected, for exactly that consistency.
  levels <- levels[nzchar(levels)]

  if (!length(levels) && length(x)) {
    abort(
      paste0(
        "`", arg, "` has no namespace levels: ",
        encodeString(paste(x, collapse = ", "), quote = "\""), "."
      ),
      call = call
    )
  }
  levels
}

#' List namespaces in a catalog
#'
#' @param catalog An `icebergr_catalog` from [icebergr_catalog()].
#' @param parent Optional parent namespace, to list only its children. Accepts
#'   `"a.b"` or `c("a", "b")`.
#'
#' @return A character vector of namespaces, dot-separated when nested.
#'
#' @examples
#' warehouse <- tempfile("warehouse")
#' dir.create(warehouse)
#' catalog <- icebergr_catalog("memory", warehouse = warehouse)
#' icebergr_list_namespaces(catalog)
#' @export
icebergr_list_namespaces <- function(catalog, parent = NULL) {
  check_catalog(catalog)
  rs_list_namespaces(catalog$ptr, as_namespace(parent, "parent", allow_null = TRUE))
}

#' List tables in a namespace
#'
#' @param catalog An `icebergr_catalog` from [icebergr_catalog()].
#' @param namespace The namespace to list. Accepts `"db"` or `c("a", "b")`.
#'
#' @return A character vector of table names, without the namespace prefix.
#'
#' @examples
#' warehouse <- tempfile("warehouse")
#' dir.create(warehouse)
#' catalog <- icebergr_catalog("memory", warehouse = warehouse)
#' # A namespace has to exist before it can hold tables.
#' icebergr_create_namespace(catalog, "db")
#' icebergr_list_tables(catalog, "db")
#' @export
icebergr_list_tables <- function(catalog, namespace) {
  check_catalog(catalog)
  rs_list_tables(catalog$ptr, as_namespace(namespace))
}

# `user:password@host` in a URI is a credential, and `uri` is a property like
# any other -- so printing it verbatim contradicted the line below and put the
# password one `saveRDS()` or knitr cache away.
#' @noRd
redact_userinfo <- function(x) {
  sub("^([A-Za-z][A-Za-z0-9+.-]*://)[^/@]*@", "\\1<redacted>@", x)
}

#' @export
print.icebergr_catalog <- function(x, ...) {
  cat("<icebergr_catalog>\n")
  cat("  type:      ", x$type, "\n", sep = "")
  cat("  name:      ", x$name, "\n", sep = "")
  if (!is.null(x$uri)) cat("  uri:       ", redact_userinfo(x$uri), "\n", sep = "")
  if (!is.null(x$warehouse)) cat("  warehouse: ", x$warehouse, "\n", sep = "")
  # Properties are deliberately not shown: they routinely hold credentials.
  invisible(x)
}
